heron/dsl/src/java/com/twitter/heron/dsl/impl/StreamletImpl.java

```diff
@@ -0,0 +1,296 @@
+   * Clones the current Streamlet. It returns an array of numClones Streamlets where each
+   * Streamlet contains all the tuples of the current Streamlet
+   * @param numClones The number of clones to clone
+   */
+  @Override
+  public List<Streamlet<R>> clone(int numClones) {
+    List<Streamlet<R>> retval = new ArrayList<>();
+    for (int i = 0; i < numClones; ++i) {
+      retval.add(repartition(getNumPartitions()));
+    }
+    return retval;
+  }
+
+  /**
+   * Returns a new Streamlet by accumulating tuples of this streamlet over a WindowConfig
+   * windowConfig and applying reduceFn on those tuples
+   * @param windowConfig This is a specification of what kind of windowing strategy you like
+   * to have. Typical windowing strategies are sliding windows and tumbling windows
+   * @param reduceFn The reduceFn to apply over the tuples accumulated on a window
+   */
+  @Override
+  public KVStreamlet<WindowInfo, R> reduceByWindow(WindowConfig windowConfig,
+                                                   SerializableBinaryOperator<R> reduceFn) {
+    ReduceByWindowStreamlet<R> retval = new ReduceByWindowStreamlet<>(this,
+                                                                      windowConfig, reduceFn);
+    addChild(retval);
+    return retval;
+  }
+
+  /**
+   * Returns a new Streamlet thats the union of this and the ‘other’ streamlet. Essentially
+   * the new streamlet will contain tuples belonging to both Streamlets
+  */
+  @Override
+  public Streamlet<R> union(Streamlet<? extends R> other) {
+    StreamletImpl<? extends R> joinee = (StreamletImpl<? extends R>) other;
+    UnionStreamlet<R> retval = new UnionStreamlet<>(this, joinee);
+    addChild(retval);
+    joinee.addChild(retval);
+    return retval;
+  }
+
+  /**
+   * Logs every element of the streamlet using String.valueOf function
+   * Note that LogStreamlet is an empty streamlet. That is its a streamlet
+   * that does not contain any tuple. Thus this function returns void.
+   */
+  @Override
+  public void log() {
+    LogStreamlet<R> logger = new LogStreamlet<>(this);
+    addChild(logger);
+    return;
+  }
+
+  /**
+   * Only used by the implementors
+   */
+  protected StreamletImpl() {
+    this.nPartitions = -1;
+    this.children = new LinkedList<>();
+    this.built = false;
+  }
+
+  public void build(TopologyBuilder bldr, Set<String> stageNames) {
+    if (built) {
+      throw new RuntimeException("Logic Error While building " + getName());
+    }
+    if (build_this(bldr, stageNames)) {
+      built = true;
+      for (StreamletImpl<?> streamlet : children) {
+        streamlet.build(bldr, stageNames);
+      }
+    }
+  }
+
+  // This is the main interface that every Streamlet implementation should implement
+  // The main tasks are generally to make sure that appropriate names/partitions are
+  // computed and add a spout/bolt to the TopologyBuilder
+  public abstract boolean build_this(TopologyBuilder bldr, Set<String> stageNames);
```

Comment:
is there a better name? java methods usually don't have underscores. Maybe doBuild()? Make this protected too since it should never be called externally?

Reply:
done